crates/bevy_ecs/src/query/state.rs

```diff
@@ -45,11 +54,14 @@ pub struct QueryState<D: QueryData, F: QueryFilter = ()> {
 
 impl<D: QueryData, F: QueryFilter> fmt::Debug for QueryState<D, F> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("QueryState")
-            .field("world_id", &self.world_id)
-            .field("matched_table_count", &self.matched_table_ids.len())
-            .field("matched_archetype_count", &self.matched_archetype_ids.len())
-            .finish_non_exhaustive()
+        let mut debug = f.debug_struct("QueryState");
+        debug.field("world_id", &self.world_id);
+        if D::IS_DENSE && F::IS_DENSE {
+            debug.field("matched_table_count", &self.matched_storage_ids.len());
+        } else {
+            debug.field("matched_archetype_count", &self.matched_storage_ids.len());
+        }
```

Comment:
Nit: could we keep the original output? Personally, I think it's a bit more readable than having to remember that the field name could change, which might be easy to miss if debugging

```suggestion
        if D::IS_DENSE && F::IS_DENSE {
            debug.field("matched_table_count", &self.matched_storage_ids.len());
            debug.field("matched_archetype_count", &0);

        } else {
            debug.field("matched_table_count", &0);
            debug.field("matched_archetype_count", &self.matched_storage_ids.len());
        }
```

Reply:
It seems odd to always be returning 0 in those cases. The way I view it, this is specializing the output on the generic type params.

Reply:
Figured out a way to avoid this.